### ISSUE TITLE
convert_data.lua: Memory leak, wrong conception or configuration?

### ISSUE THREAD

**Issue by JSkalskiSBG** [nagadomi/waifu2x#218]
I have 100.000 images for 'training' (around 1MB each). I ran:

    find big_images/ -name "*.png" > data/image_list.txt
    th convert_data.lua -use_transparent_png 1

and it crashed after 40.000 images (PC: 16GB ram, 16GB swap).

After crash I've noticed that nothing saved to data/images.t7.
Problem is RAM usage. It looks like it generates whole 'data/images.t7' in RAM memory and dump it to file after processing all images.

**Is it really required? Can't it write processed images to file on HDD?**

Did I configure something wrong or is it a wrong conception in waifu?
------------------------
I wrote short bash to split image_data.txt to 13 files and load it one by one and then rename generated images.7t, but there should be solution in waifu2x config/code.

**Comment by nagadomi** [comment]
convert_lua.lua/train.lua loads all the images in memory compressed with snappy.
If the RAM is small, you need to reduce the number of images.

>Is it really required?

No. HDD i/o and PNG decompression is slow, but can be processed in multi processing. The current code does not support it.

And, if you are going to try train, I recommend using the `dev` branch. It has improved train performance.

**Comment by JSkalskiSBG** [comment]
I wrote bash scripts to split training into 2000 images per train data set and learn with these sets.
I left train script for 60 hours, but results are worse then 'anime art' results.
http://vpn.skalski.pro/items_nn.htm
(last image is my network 8x scale image, pre-last image is anime-art network 8x scale image, first image is source image)

@nagadomi 
My network produces some artifacts. Light things become totally white and **there are some hashes '#' on images.**
1. There was a thread about this problem with 'anime' network, but you somehow solved it. How did you do that?
2. Is there any official forum for waifu2x users where they can share their problems/experience? Github issues isn't best place for that.

----------------------

## How to train network with high number of images without enough RAM to store all images in it:

1. Remove everything from 'data' folder in 'waifu2x'. Your images should be in folder 'my_images' with extension '.png'. 'test file' to check quality of NN results should be in file 'test.png'.
2. Create file **split_convert.sh** and paste:

```
    #!/bin/bash
    FILES=data/imagessplit*
    for f in $FILES
    do
        echo "Processing $f file..."
        cp $f "data/image_list.txt"
        th convert_data.lua -use_transparent_png 1
        cp "data/images.t7" $f".t7"
    done
```

3. Create file **split_train.sh** and paste  (in this file you can adjust learning rate):

```
    #!/bin/bash
    FILES=data/imagessplit*.t7
    lr=1000000
    while true;
    do
        for f in $FILES
        do
          echo "Processing $f file"
          echo "0.0000"$lr
          cp $f "data/images.t7"
          th train.lua -model_dir models/my_model -method scale -scale 2 -test test.png -use_transparent_png 1 -epoch 1 -inner_epoch 3 -learning_rate "0.0000"$lr -resume models/my_model/scale2.0x_model.t7
        done
        lr=$((lr * 95 / 100))
    done
```
4. Execute in console (in waifu2x folder) - train with PNGs with transparency:

```
	find my_images/ -name "*.png" > data/image_list.txt
	split -l 2000 data/image_list.txt --numeric-suffixes data/imagessplit
	sh split_convert.sh
	sh split_train.sh
```
Your 'network' will be stored in 'models/my_model/scale2.0x_model.t7'.
